Manage downstream datasets with locationless source dataset

santoshamohan opened the pull request:
# Reason for this pull request
Datasets within the production database had multiple `derived datasets` / `duplicate datasets` derived from the source dataset whose `local_uri` is set to `None`/`Not Available`. This was causing ingestion failure during orchestration process. We need a new tool to manage this scenario and take appropriate action.

**Background**
The above mentioned scenario could arise if we delete the file from the disk and run the `sync` tool with `--update-location` option. Dataset location within the database are set to `None`/`Not Available` by the `sync` tool. So, during `ingest` process, a query from the datacube would result in `index out of range` error and storage file creation fails.

# Proposed solutions
Following are the changes implemented as part of this PR.
1. Fixed `dea-duplicates`
    -  Required changing the setup in `collections.py` for unique fields
    -  Updated the tests since the `CSV` output depends on the unique fields
2. Added `NBART` scenes to `collections.py`
3. Fixed the output directory for the `dea-clean` archived tool
4. Add a new feature to list all the `downstream` datasets whose source dataset has no location
5. Extend `dea-coherence` to check `downstream` datasets
6. Log erroneous datasets to a `CSV` file

- [x] Tests added

Commit:
Add command line option to check bad downstream datasets

Commit:
Remove blank line containing  whitespace

Commit:
Add tests to test dea-coherence

Commit:
Update tests for dea-coherence

Commit:
Fix pylint error

Commit:
Revert collection name to ls8_scene_test

Commit:
Revert changes to conftest.py

santoshamohan committed:
Merge branch 'develop' into downstream-datasets

Commit:
Update references to local_uri

Commit:
Update log info

Commit:
Ignore check-siblings for dsm1sv10

Commit:
Ignore process/check siblings for nbar/nbart/pq albers products

santoshamohan committed:
Update comment

Commit:
Log erroneous datasets to the csv file

santoshamohan committed:
Merge branch 'develop' into downstream-datasets

omad commented on digitalearthau/coherence.py:
```diff
@@ -1,17 +1,45 @@
+_LOG = structlog.getLogger('dea-coherence')
+_dataset_count, _siblings_count, _locationless_count, _archive_count = 0, 0, 0, 0
+_downstream_dataset_count, _downstream_summary_prod_count = 0, 0
+_bad_downstream_dataset = collections.deque({})
```
It's not necessary to use a `deque` here, it will work, but so will an ordinary list would be fine.

Using `deque` makes me think that this is used for some special or performance purpose... but it isn't.

Also, it will initialise to empty by default, so the extra `{}` is doubly unnecessary. :)

santoshamohan replied:
Changed `deque` to a simple `set`.

omad commented on digitalearthau/coherence.py:
```diff
@@ -1,17 +1,45 @@
+
+# Product list containing Telemetry, Level1, Level2, NBAR, PQ, PQ_Legacy, WOfS, FC products
+# Level1_Scene datasets are expected not to have any locations (in future)
+PRODUCT_TYPE_LIST = ['ls5_satellite_telemetry_data', 'ls5_level1_scene', 'ls5_nbar_scene', 'ls5_nbart_scene',
```
We should be able to get this list from `digitalearthau.collections`, it'll be too easy for it to fall out of date if we need to list things in more than one place.

santoshamohan replied:
Replaced hard-coded values of `product` names to `digitalearthau.collections.registered_collection_names()`

omad commented on digitalearthau/coherence.py:
```diff
@@ -1,17 +1,45 @@
 
 
-@click.command()
+@click.group(help=__doc__)
```
If there's only the one command, we don't need to have a `click.group`

santoshamohan replied:
Removed `click.group`

omad commented on digitalearthau/coherence.py:
```diff
@@ -28,6 +56,10 @@
+@click.option('--check-downstream-ds',
+              is_flag=True,
+              default=False,
+              help='Check if downstream datasets have their ancestor datasets archived')
```
Make it clear whether this check is for *archived parents* or *locationless parents* or *both*.

santoshamohan replied:
Click option `help` string updated.

omad commented on digitalearthau/coherence.py:
```diff
@@ -111,24 +166,24 @@ def _archive_duplicate_siblings(dc, ids):
             if source_dataset.is_archived:
                 _LOG.info(
-                    "ancestor.dataset_id",
+                    "ancestor." + str(dataset.type.name) + ".dry_run",
```
Feel free to use the new Python 3.6 style f-strings in `digitalearthau`.
This can be simplified to `f"ancestor.{dataset.type.name}.dry_run"`

santoshamohan replied:
Modified to python 3.6 style f-strings.

omad commented on digitalearthau/coherence.py:
```diff
@@ -139,18 +194,80 @@ def _check_ancestors(check_ancestors: bool,
-                if siblings:
+
+                # Do not process/check siblings for nbar/nbart/pq albers products
+                if siblings and (str(dataset.type.name) not in IGNORE_SIBLINGS):
```
The `(str(dataset.type.name) not in IGNORE_SIBLINGS)` is redundant, it's already checked on line 185.

santoshamohan replied:
`--check-siblings` command line option removed as this is handled by `--check-downstream` option and by `dea-duplicates` command.

omad commented on digitalearthau/coherence.py:
```diff
@@ -139,18 +194,80 @@ def _check_ancestors(check_ancestors: bool,
+
+def _fetch_derived_datasets(dataset, dc):
+    to_process = {dataset.id}
+    derived_dataset = collections.deque({})
```
This would be better as `derived_datasets = set()`, and then using `derived_datasets.update(derived)` below will give you a single set of all the derived datasets, instead of the current collection of collections, leading to the nested `for` loops on line 223-4 and 242-3.

santoshamohan replied:
Updated to remove any discrepancies.

omad commented on digitalearthau/coherence.py:
```diff
@@ -139,18 +194,80 @@ def _check_ancestors(check_ancestors: bool,
+    """
+
+    # Fetch all the derived datasets using the source dataset.
+    derived_dataset = _fetch_derived_datasets(dataset, dc)
```
It looks like `_fetch_derived_datasets()` is being called twice (here and line 230), but I think it only needs to be called once. It already walks down the entire tree of child datasets.

santoshamohan replied:
`_fetch_derived_datasets()` is first called with dataset to fetch all derived datasets. Then in the second call, the same function is called with locationless/archived dataset to fetch all the derived datasets of the same.

omad reviewed (changes requested):
Hey Santosh, thanks for getting this working.

There's a few things here that it'd be good to clean up before merging. Some are really easy fixes, but some of the logic is also getting hard to reason about and we should sit down and work through it together to see if we can simplify it.

Commit:
Update as per review comments

Commit:
Add nbart to registered collection names and fix unique field in collection

Commit:
Fix integration test scripts

Commit:
Fix integration test scripts

Commit:
Fix travis failures

Commit:
Fix travis py.test file location issue

Commit:
Remove redundant line of code

omad committed:
Reformatting and cleaning

- `global` is only required when assigning to a global variable
- The product list only needs to be computed once
- Change argument ordering to be consistent between functions

omad committed:
More coherence updates, incomplete

Commit:
More refactoring of coherence logic

Commit:
Update dea-coherence command option help string

Commit:
Update integration test scripts

Commit:
Fix py.test failures

Commit:
Update pytest script to fix codecov/patch failure

Commit:
Update pytest script to fix codecov/patch failure

Commit:
Appease travis checks

Commit:
Appease travis yamllint checks

Commit:
Fix pytest failures in full sync test script

omad committed:
Tiny changes, trying to make more grokkable

omad committed:
Readability refactoring of coherence

omad committed:
Refactor coherence boolean logic

omad committed:
Organise imports and fix deprecation warnings

santoshamohan committed:
Merge branch 'develop' into downstream-datasets

santoshamohan committed:
Merge pull request #188 from GeoscienceAustralia/damien/downstream-datasets

More coherence check improvements

omad commented on integration_tests/conftest.py:
```diff
@@ -237,6 +255,38 @@ def other_dataset(integration_test_data: Path, test_dataset: DatasetForTests) ->
 
+@pytest.fixture
+def ls8_pq_scene_test_dataset(integration_test_data, dea_index) -> DatasetForTests:
+    """A dataset on disk, with corresponding collection"""
```
What's the purpose of this fixture, and how is it different to `test_dataset`?

santoshamohan replied:
`test_dataset` contains `level1` dataset that would be skipped by the `coherence` tool. Hence, I had to add new `pq_scene` fixture.

omad commented on integration_tests/conftest.py:
```diff
@@ -237,6 +255,38 @@ def other_dataset(integration_test_data: Path, test_dataset: DatasetForTests) ->
+    test_data = integration_test_data
+
+    ls8_collection = Collection(
+        name='ls8_scene_test',
```
I think this should be different, this is a PQ dataset...

santoshamohan replied:
Agreed. Updated the collection name to be `ls8_pq_scene_test`.

omad commented on integration_tests/conftest.py:
```diff
@@ -237,6 +255,38 @@ def other_dataset(integration_test_data: Path, test_dataset: DatasetForTests) ->
+    ls8_collection = Collection(
+        name='ls8_scene_test',
+        query={},
+        file_patterns=[str(test_data.joinpath('LS8*/ga-metadata.yaml'))],
```
This will match more than the newly added PQ dataset.

santoshamohan replied:
Updated the directory name to match only the newly added `PQ` dataset.

omad commented on integration_tests/test_coherence.py:
```diff
@@ -0,0 +1,147 @@
+
+
+def test_check_locationless(ls8_pq_scene_test_dataset: DatasetForTests,
+                            other_dataset: DatasetForTests,
```
It's clear what this test is supposed to be checking, but I don't think the flow is clearly enough explained.

santoshamohan replied:
Added more comments to aid the test flow.

omad commented on integration_tests/test_coherence.py:
```diff
@@ -0,0 +1,147 @@
+    timerange = ["2000-01-01 < time < 2018-12-31"]
+    exe_opts.extend(timerange)
+
+    assert_click_command(coherence.main, exe_opts)
```
This should check for correct output, not only that the coherence tool ran without errors

santoshamohan replied:
Updated test to check the expected output.

omad commented on integration_tests/test_full_sync.py:
```diff
@@ -26,7 +26,8 @@
                              integration_test_data: Path,
-                             other_dataset: DatasetForTests):
+                             other_dataset: DatasetForTests,
+                             ls8_pq_scene_test_dataset: DatasetForTests):
```
I don't think that `ls8_pq_scene_test_dataset` should be added to these tests. Maybe it was getting pulled in due to the fixtures not setting up the `Collections` correctly, as I've pointed out above.

santoshamohan replied:
Reverted back the changes and updated `conftest.py` to pull correct `collections`.

omad reviewed (dismissed):
Hi Santosh, I think this is looking much better. I think there's still a few issues with the changes in tests.

Also, please try to list all the changes in a PR, for example for this PR:
-   Fixed `dea-duplicates`
    -   Required changing the setup in `collections.py` for unique fields
    -   Updated the **tests** since the CSV output depends on the unique fields
-   Added **NBART** scenes to `collections.py`
-   Fixed the output directory for the `dea-clean archived` tool
-   Extend `dea-coherence` to check downstream datasets

Commit:
Refactor coherence test script

Commit:
Print coherence summary using click command instead of structlog way

santoshamohan committed:
Merge branch 'develop' into downstream-datasets

Commit:
Fix clirunner output typo mismatch

santoshamohan committed:
Merge branch 'develop' into downstream-datasets

santoshamohan committed:
Merge branch 'develop' into downstream-datasets

santoshamohan committed:
Merge branch 'develop' into downstream-datasets

Commit:
Update doc string

santoshamohan committed:
Merge branch 'develop' into downstream-datasets

santoshamohan committed:
Merge branch 'develop' into downstream-datasets

Commit:
Add barest earth and s2 products to the collections

omad committed:
Merge branch 'develop' into downstream-datasets
